lib/llm-events/langchain/event.js

```diff
@@ -0,0 +1,94 @@
+/*
+ * Copyright 2024 New Relic Corporation. All rights reserved.
+ * SPDX-License-Identifier: Apache-2.0
+ */
+
+'use strict'
+
+const { makeId } = require('../../util/hashes')
+const { DESTINATIONS } = require('../../../lib/config/attribute-filter')
+const { isSimpleObject } = require('../../util/objects')
+
+/**
+ * @typedef {object} LangChainEventParams
+ * @property {object} agent A New Relic agent instance.
+ * @property {object} segment A New Relic segment instance.
+ * @property {string} runId The identifier LangChain has assigned to the "run."
+ * @property {Object<string, string>} metadata The metadata, if any, associated with the
+ * LangChain run.
+ * @property {boolean|undefined} [virtual=true] Indicates that this event is a
+ * LangChain specific event (`true`). LangChain is not itself an LLM, but an
+ * interface to many LLMs. Any LLMs LangChain interacts with that we have
+ * instrumented will have their own traces that are not "virtual."
+ */
+/**
+ * @type {LangChainEventParams}
+ */
+const defaultParams = {
+  agent: {},
+  segment: {
+    transaction: {}
+  },
+  runId: '',
+  metadata: {},
+  virtual: undefined
+}
+
+/**
+ * Baseline object representing an event in a LangChain conversation.
+ */
+class LangChainEvent {
+  id = makeId(36)
+  appName
+  conversation_id
+  span_id
+  request_id
+  transaction_id
+  trace_id
+  ingest_source = 'Node'
+  vendor = 'langchain'
+  virtual_llm = true
+
+  #metadata = {}
+
+  constructor(params = defaultParams) {
+    params = Object.assign({}, defaultParams, params)
+    const { agent, segment } = params
+
+    this.appName = agent.config.applications()[0]
+    this.conversation_id = this.conversationId(agent)
+    this.span_id = segment?.id
+    this.request_id = params.runId
+    this.transaction_id = segment?.transaction?.id
+    this.trace_id = segment?.transaction?.traceId
+    this.metadata = params.metadata
+
+    if (params.virtual !== undefined) {
+      if (params.virtual !== true && params.virtual !== false) {
+        throw Error('params.virtual must be a primitive boolean')
+      }
+      this.virtual_llm = params.virtual
+    }
+  }
+
+  get metadata() {
+    return this.#metadata
+  }
+
+  set metadata(value) {
+    if (isSimpleObject(value) === false) {
+      return
+    }
+    for (const [key, val] of Object.entries(value)) {
+      this.#metadata[key] = val
+    }
+  }
+
+  conversationId(agent) {
```

Comment:
any chance we can abstract this? we have the same logic in llm-events/openai/event.js

Reply:
I think that would be adding complexity for not much benefit. If we need to add another custom event class that implements this same logic, then I would agree that it should be revisited.

Reply:
we already do have this in [here](https://github.com/newrelic/node-newrelic/blob/main/lib/llm-events/openai/event.js#L51)

Reply:
Right. That's the one other place. I ascribe to the write it three times then generalize it rule.